### PR TITLE
init feature proposal 2 process

### LINKED ISSUE #17: Feature Proposal #2 Apartment Experiment
Create an elevator (1 dimension) world with characters and a future event list,

some features I am thinking of,

characters will have a state, for idling, or speaking (with another character), eating, sleeping,

characters can move up and down the apartment at random,

if they're in a room with other characters, they might have a chance to engage speaking with another character,

and at that point, both the characters will have their state be speaking, provided they're both idling before,

characters who are eating and sleeping cannot be spoken with,

keep track of a world history, with the event's time as key, 

when a character decide to speak with another character, they create a speaking event,

the order of speaking event triggering matters, the first event in the future event list is triggered first,

when the event is triggered, the state of both the characters in question, if they were both idling before, is immediately set to speaking,

if both characters are successfully speaking, a new  end speaking event will be created, at a random time in the future,

and when that end speaking event's time key is reached by the simulation time, and that event is popped from the future event list,

the characters are set to idling again,

characters who are also idling can have a chance to randomly create speaking, eating, and sleeping events

### PR DESCRIPTION
**Pull request by youhengzhou**
will close #17 when it finishes

Create an elevator (1 dimension) world with characters and a future event list,

some features I am thinking of,

characters will have a state, for idling, or speaking (with another character), eating, sleeping,

characters can move up and down the apartment at random,

if they're in a room with other characters, they might have a chance to engage speaking with another character,

and at that point, both the characters will have their state be speaking, provided they're both idling before,

characters who are eating and sleeping cannot be spoken with,

keep track of a world history, with the event's time as key,

when a character decide to speak with another character, they create a speaking event,

the order of speaking event triggering matters, the first event in the future event list is triggered first,

when the event is triggered, the state of both the characters in question, if they were both idling before, is immediately set to speaking,

if both characters are successfully speaking, a new end speaking event will be created, at a random time in the future,

and when that end speaking event's time key is reached by the simulation time, and that event is popped from the future event list,

the characters are set to idling again,

characters who are also idling can have a chance to randomly create speaking, eating, and sleeping events

### CODE REVIEW

File: base.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+jdb = jsoneng.JsonDB()
+jdb.create({})
+
+import random
+import heapq
+
+class Character:
+    def __init__(self, name):
+        self.name = name
+        self.state = "idling"
```
Comment (1112zakaria) [inline base.py:16]: I prefer the use of enums instead of strings for representing state.

File: base.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    def __repr__(self):
+        return self.name
+
+class Event:
+    def __init__(self, time, action, participants=None):
+        self.time = time
+        self.action = action
+        self.participants = participants if participants is not None else []
```
Comment (1112zakaria) [inline base.py:25]: Enum for actions perhaps?

File: ApartmentBuildingSim.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    def __init__(self, time, chars, name):
+        self.time = time
+        self.name = name
+        self.chars = chars
+
+    def __repr__(self):
+        return f"EVENT: time:{self.time} chars:{self.chars[0].name} name:{self.name}"
+
+    def __lt__(self, other):
+        return self.time < other.time
```
Comment (1112zakaria) [inline ApartmentBuildingSim.py:42]: Suggestion: define method `handleEvent()` in Event but throw an exception when it is called. This is to communicate to developers that this is an abstract class. Could be considered extra unneeded boilerplate though...
image

```
   def handleEvent(self, futureEventList):
      raise NotImplementedError()
```

File: ApartmentBuildingSim.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            c.state = 'stage 0: pissing'
+        heapq.heappush(futureEventList, PissWater1(self.time + random.randint(4,8), self.chars))
+
+class PissWater1(Event):
+    def __init__(self, time, chars):
+        super().__init__(time, chars, 'stage 1: finish pissing')
+
+    def handleEvent(self, futureEventList):
+        for c in self.chars:
+            c.state = 'idle'
```
Comment (1112zakaria) [inline ApartmentBuildingSim.py]: Lots of repeat looping `for c in self.chars`. 
Suggestion: add method `setCharacters(state, action) in class Event
```
class Event:
   ...
   def setCharacters(self, state=None, action=None):
      for c in self.chars:
         if state:
            c.state = state
         if action:
            c.action = action

...

class InitCharsEvent
   def handleEvent:
      self.setCharacters('idle', 'lobby')
 
```

File: ApartmentBuildingSim.py (hunk 3 above)
Comment (anishtankala) [reply ApartmentBuildingSim.py:42]: good suggestion!

File: base.py (hunk 2 above)
Comment (youhengzhou) [reply base.py:25]: it should be location actually but oh this is actually not the right file

but in the future we'll do classes for those as enums, dataclasses instead

File: base.py (hunk 1 above)
Comment (youhengzhou) [reply base.py:16]: we'll do it in the future with the next sprint!

**Pull request merged**